util/syspolicy/setting/snapshot.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package setting
+
+import (
+	"slices"
+	"strings"
+
+	xmaps "golang.org/x/exp/maps"
+	"tailscale.com/util/deephash"
+)
+
+// Snapshot is an immutable collection of ([Key], [RawItem]) pairs, representing
+// a set of policy settings applied at a specific moment in time.
+// A nil pointer to [Snapshot] is valid.
+type Snapshot struct {
+	m       map[Key]RawItem
+	sig     deephash.Sum // of m
+	summary Summary
+}
+
+// NewSnapshot returns a new [Snapshot] with the specified items and options.
+func NewSnapshot(items map[Key]RawItem, opts ...SummaryOption) *Snapshot {
+	return &Snapshot{m: items, sig: deephash.Hash(&items), summary: SummaryWith(opts...)}
+}
+
+type keyItemPair struct {
+	Key  Key
+	Item RawItem
+}
+
+// All returns an iterator over [[Key], [RawItem]] key-value pairs in b. The
+// iteration order is not specified and is not guaranteed to be the same from
+// one call to the next.
+func (s *Snapshot) All() []keyItemPair {
+	if s == nil {
+		return nil
+	}
+	// TODO(nickkhyl): return iter.Seq2[[Key], [RawItem]] in Go 1.23,
+	// and remove [keyItemPair].
+	items := make([]keyItemPair, 0, len(s.m))
+	for k, i := range s.m {
+		items = append(items, keyItemPair{k, i})
+	}
+	return items
+}
+
+// Get returns the value of the policy setting with the specified key
+// or nil if it does not exist or could not be read.
+func (s *Snapshot) Get(k Key) any {
+	v, _ := s.GetErr(k)
+	return v
+}
+
+// GetErr returns the value of the policy setting with the specified key,
+// [ErrNotConfigured] if it does not exist, or an error returned by
+// the policy Store if the policy setting could not be read.
+func (s *Snapshot) GetErr(k Key) (any, error) {
+	if s != nil {
+		if s, ok := s.m[k]; ok {
+			return s.Value(), s.Error()
+		}
+	}
+	return nil, ErrNotConfigured
+}
+
+// GetSetting returns the untyped policy setting with the specified key and true
+// if a policy setting with such key has been configured;
+// otherwise, it returns zero, false.
+func (s *Snapshot) GetSetting(k Key) (setting RawItem, ok bool) {
+	setting, ok = s.m[k]
+	return setting, ok
+}
+
+// Equal reports whether s and s2 are equal.
+func (s *Snapshot) Equal(s2 *Snapshot) bool {
+	if !s.EqualItems(s2) {
+		return false
+	}
+	return s.Summary() == s2.Summary()
+}
+
+// EqualItems reports whether items in s and s2 are equal.
+func (s *Snapshot) EqualItems(s2 *Snapshot) bool {
+	if s == s2 {
+		return true
+	}
+	if s.Len() != s2.Len() {
+		return false
+	}
+	if s.Len() == 0 {
+		return true
+	}
+	return s.sig == s2.sig
```

Comment:
How fast does this need to be? `maps.Equal` may suffice and then there's no need to use `deephash`.

Reply:
I needed the hash in `sig` for something else initially, and decided to use it for equality comparison as well. But I ended up removing that initial thing, and there's really no reason to keep it `deephash` just here.
Will remove and use `maps.Equal`, thanks!

Reply:
I decided to revert this change. 

`RawItem` may contain a value of an incomparable type, such as `[]string`, and is therefore incomparable. It also has pointer fields that should not be compared directly.

We could implement `RawItem.Equal` and then use it with `maps.EqualFunc`, or we could use `reflect.DeepEqual` to compare the maps. But I'm not sure it's worth it since `deephash` is already imported by `tailscaled` and other packages that have legitimate reasons to use `syspolicy`. And we should probably split `ipn` into two packages if we'd like to remove unnecessary dependencies, including a dependency on `syspolicy`, from `derper` and `tailscale` (the CLI).

Reply:
I feel like if the alternative is `reflect.DeepEqual` then that's not going to be much better. Ideally we should figure out how to not need to do either of those things, but maybe we'll have some better ideas after getting experience using this. It's fine for now.